connection_pool/connection_pool_test.go

```diff
@@ -233,6 +234,223 @@ func TestClose(t *testing.T) {
 	require.Nil(t, err)
 }
 
+type testListener struct {
+	added, closed int
+	errs          []error
+	mutex         sync.Mutex
+}
+
+func (tl *testListener) addErr(err error) {
+	tl.errs = append(tl.errs, err)
+}
+
+func (tl *testListener) Add(conn *tarantool.Connection, role connection_pool.Role) error {
+	tl.mutex.Lock()
+	defer tl.mutex.Unlock()
+
+	tl.added++
+
+	if conn == nil {
+		tl.addErr(fmt.Errorf("added conn == nil"))
+		return nil
+	}
+
+	addr := conn.Addr()
+	if addr == servers[0] {
+		if role != connection_pool.MasterRole && tl.closed != 1 ||
+			role != connection_pool.ReplicaRole && tl.closed == 1 {
+			tl.addErr(fmt.Errorf("unexpected added role %d for addr %s", role, addr))
+		}
+	} else if addr == servers[1] {
+		if role != connection_pool.ReplicaRole {
+			tl.addErr(fmt.Errorf("unexpected added role %d for addr %s", role, addr))
+		}
+	} else {
+		tl.addErr(fmt.Errorf("unexpected added addr %s", addr))
+	}
+
+	return nil
+}
+
+func (tl *testListener) Removed(conn *tarantool.Connection, role connection_pool.Role) {
```

Comment:
So `Add` may return error yet `Removed` is not?

Reply:
Thank you for that excellent point. This may be useful in the future. After update `Closed` (renamed `Removed`) returns an error too.